src/compiler/scala/tools/nsc/backend/jvm/BCodeBodyBuilder.scala

```diff
@@ -310,6 +310,20 @@ abstract class BCodeBodyBuilder extends BCodeSkelBuilder {
         case app : Apply =>
           generatedType = genApply(app, expectedType)
 
+        case app @ ApplyDynamic(qual, Literal(Constant(reflectiveMethodType: MethodType)) :: Nil) if qual.symbol == MethodCacheIndy =>
+          val returnAsmType = typeToBType(qual.symbol.info.resultType).toASMType
+          val invokedType = asm.Type.getMethodType(returnAsmType)
+          val reflectiveParams = reflectiveMethodType.params.map(p => typeToBType(p.info))
+          val reflectiveParamsDescriptor = MethodBType(reflectiveParams, ObjectRef).toASMType.getDescriptor
+          mnode.visitInvokeDynamicInsn(qual.symbol.name.encoded, invokedType.getDescriptor, structuralCallSite_BoostrapHandle, reflectiveParamsDescriptor)
+
+        case app @ ApplyDynamic(qual, Literal(Constant(symname: String)) :: Nil) if qual.symbol == Symbol_apply =>
+          val symbol = app.symbol
+          val returnAsmType = typeToBType(symbol.info.resultType).toASMType
+          val invokedType = asm.Type.getMethodType(returnAsmType)
+          val name = "apply"
+          mnode.visitInvokeDynamicInsn(name, invokedType.getDescriptor, symbolLiteralBoostrapHandle, symname)
+
```

Comment:
This is the ad-hoc part I was referring to; ideally the `ApplyDynamic` tree would have all the info we need to emit the `invokedynamic` instruction.